File: go/internal/anomaly/anomaly.go
// Package anomaly is the statistical/pattern anomaly layer. It watches each agent's recent
// behaviour and flags loops (the same tool hammered over and over), rate spikes, and
// cyclic A-B-A-B patterns. It is deterministic and cheap - it runs on every event before
// any LLM is consulted.
package anomaly

import (
	"fmt"
	"sync"

	"github.com/parag-labs/guardianforge/go/internal/models"
)

// Detector keeps a bounded per-agent history and scores anomalies.
type Detector struct {
	mu      sync.Mutex
	history map[string][]string
	window  int
	// loopThreshold: how many identical calls in the window trips a loop.
	loopThreshold int
	// rateThreshold: how many events in the window trips a rate spike.
	rateThreshold int
}

// New builds a detector with sensible defaults.
func New() *Detector {
	return &Detector{
		history:       map[string][]string{},
		window:        12,
		loopThreshold: 5,
		rateThreshold: 10,
	}
}

// fingerprint of an event for pattern matching.
func fp(ev models.AgentEvent) string {
	return string(ev.Type) + ":" + ev.Tool + ":" + ev.Action
}

// Observe records an event and returns any anomalies detected for the agent.
func (d *Detector) Observe(ev models.AgentEvent) []models.Anomaly {
	d.mu.Lock()
	defer d.mu.Unlock()

	h := append(d.history[ev.AgentID], fp(ev))
	if len(h) > d.window {
		h = h[len(h)-d.window:]
	}
	d.history[ev.AgentID] = h

	var out []models.Anomaly
	if a, ok := d.detectLoop(ev.AgentID, h); ok {
		out = append(out, a)
	}
	if a, ok := d.detectRate(ev.AgentID, h); ok {
		out = append(out, a)
	}
	if a, ok := d.detectCyclic(ev.AgentID, h); ok {
		out = append(out, a)
	}
	return out
}

// detectLoop flags many identical fingerprints among the most recent events.
func (d *Detector) detectLoop(agent string, h []string) (models.Anomaly, bool) {
	if len(h) == 0 {
		return models.Anomaly{}, false
	}
	last := h[len(h)-1]
	count := 0
	for i := len(h) - 1; i >= 0 && h[i] == last; i-- {
		count++
	}
	if count >= d.loopThreshold {
		score := clamp(float64(count) / float64(d.window))
		return models.Anomaly{
			AgentID: agent, Type: models.AnomalyLoop, Score: round(score),
			Detail: fmt.Sprintf("%d consecutive identical actions (%s)", count, last),
		}, true
	}
	return models.Anomaly{}, false
}

// detectRate flags a burst: the window is (nearly) full of events for one agent.
func (d *Detector) detectRate(agent string, h []string) (models.Anomaly, bool) {
	if len(h) >= d.rateThreshold {
		score := clamp(float64(len(h)) / float64(d.window))
		return models.Anomaly{
			AgentID: agent, Type: models.AnomalyRateSpike, Score: round(score),
			Detail: fmt.Sprintf("%d events within the observation window", len(h)),
		}, true
	}
	return models.Anomaly{}, false
}

// detectCyclic flags an alternating A-B-A-B pattern over the last several events.
func (d *Detector) detectCyclic(agent string, h []string) (models.Anomaly, bool) {
	if len(h) < 6 {
		return models.Anomaly{}, false
	}
	tail := h[len(h)-6:]
	a, b := tail[0], tail[1]
	if a == b {
		return models.Anomaly{}, false
	}
	for i, x := range tail {
		want := a
		if i%2 == 1 {
			want = b
		}
		if x != want {
			return models.Anomaly{}, false
		}
	}
	return models.Anomaly{
		AgentID: agent, Type: models.AnomalyCyclic, Score: 0.7,
		Detail: fmt.Sprintf("cyclic pattern between %q and %q", a, b),
	}, true
}

func clamp(f float64) float64 {
	if f > 1 {
		return 1
	}
	if f < 0 {
		return 0
	}
	return f
}

func round(f float64) float64 { return float64(int(f*100+0.5)) / 100 }

File: rust/src/anomaly.rs
//! Deterministic statistical/pattern anomaly detection, ported from Go `internal/anomaly`.
//!
//! It watches each agent's recent behaviour and flags loops (the same tool hammered over
//! and over), rate spikes, and cyclic A-B-A-B patterns. It is deterministic and cheap, and
//! runs on every event before any model is consulted.

use std::collections::HashMap;

use crate::models::{self, AgentEvent, Anomaly};

const WINDOW: usize = 12;
const LOOP_THRESHOLD: usize = 5;
const RATE_THRESHOLD: usize = 10;

fn clamp(f: f64) -> f64 {
    f.clamp(0.0, 1.0)
}

/// Reproduce Go's `float64(int(f*100+0.5)) / 100` (round half up for `f >= 0`).
fn round(f: f64) -> f64 {
    ((f * 100.0 + 0.5) as i64) as f64 / 100.0
}

fn fingerprint(ev: &AgentEvent) -> String {
    format!("{}:{}:{}", ev.r#type, ev.tool, ev.action)
}

/// Keeps a bounded per-agent history and scores anomalies.
#[derive(Debug, Default)]
pub struct Detector {
    history: HashMap<String, Vec<String>>,
}

impl Detector {
    /// Create an empty detector.
    pub fn new() -> Self {
        Detector {
            history: HashMap::new(),
        }
    }

    /// Record an event and return any anomalies detected for the agent.
    pub fn observe(&mut self, ev: &AgentEvent) -> Vec<Anomaly> {
        let entry = self.history.entry(ev.agent_id.clone()).or_default();
        entry.push(fingerprint(ev));
        if entry.len() > WINDOW {
            let start = entry.len() - WINDOW;
            entry.drain(0..start);
        }
        let h = entry.clone();

        let mut out = Vec::new();
        if let Some(a) = detect_loop(&ev.agent_id, &h) {
            out.push(a);
        }
        if let Some(a) = detect_rate(&ev.agent_id, &h) {
            out.push(a);
        }
        if let Some(a) = detect_cyclic(&ev.agent_id, &h) {
            out.push(a);
        }
        out
    }
}

fn detect_loop(agent: &str, h: &[String]) -> Option<Anomaly> {
    let last = h.last()?;
    let mut count = 0usize;
    for x in h.iter().rev() {
        if x == last {
            count += 1;
        } else {
            break;
        }
    }
    if count >= LOOP_THRESHOLD {
        let score = clamp(count as f64 / WINDOW as f64);
        return Some(Anomaly {
            agent_id: agent.to_string(),
            r#type: models::ANOMALY_LOOP.to_string(),
            score: round(score),
            detail: format!("{count} consecutive identical actions ({last})"),
        });
    }
    None
}

fn detect_rate(agent: &str, h: &[String]) -> Option<Anomaly> {
    if h.len() >= RATE_THRESHOLD {
        let score = clamp(h.len() as f64 / WINDOW as f64);
        return Some(Anomaly {
            agent_id: agent.to_string(),
            r#type: models::ANOMALY_RATE_SPIKE.to_string(),
            score: round(score),
            detail: format!("{} events within the observation window", h.len()),
        });
    }
    None
}

fn detect_cyclic(agent: &str, h: &[String]) -> Option<Anomaly> {
    if h.len() < 6 {
        return None;
    }
    let tail = &h[h.len() - 6..];
    let a = &tail[0];
    let b = &tail[1];
    if a == b {
        return None;
    }
    for (i, x) in tail.iter().enumerate() {
        let want = if i % 2 == 0 { a } else { b };
        if x != want {
            return None;
        }
    }
    Some(Anomaly {
        agent_id: agent.to_string(),
        r#type: models::ANOMALY_CYCLIC.to_string(),
        score: 0.7,
        detail: format!("cyclic pattern between \"{a}\" and \"{b}\""),
    })
}
